File: Leetcode/tree/113_PathSumII/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(res, new LinkedList<Integer>(), root, sum);
        return res;
    }
        private void dfs(List<List<Integer>> res, List<Integer> temp,
                        TreeNode root, int target){
            if(root==null) return;
            if(root.left==null && root.right==null){
                if(target==root.val){
                    temp.add(root.val);
                    res.add(new ArrayList(temp));
                    if(temp.size()>0) temp.remove(temp.size()-1);
                }
                return;
            }
            temp.add(root.val);
            dfs(res, temp, root.left, target-root.val);
            dfs(res, temp, root.right, target-root.val);
            if(temp.size()>0) temp.remove(temp.size()-1);
        }
}

File: Leetcode/tree/113_PathSumII/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    void dfs(vector<vector<int>> &res, vector<int>& temp,
                        TreeNode* root, int target){
            if(root==nullptr) return;
            if(root->left==nullptr && root->right==nullptr){
                if(target==root->val){
                    temp.push_back(root->val);
                    res.push_back(vector(temp));
                    if(temp.size()>0) temp.pop_back();
                }
                return;
            }
            temp.push_back(root->val);
            dfs(res, temp, root->left, target-root->val);
            dfs(res, temp, root->right, target-root->val);
            if(temp.size()>0) temp.pop_back();
        }
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        vector<int> temp;
        dfs(res, temp, root, sum);
        return res;
    }
};
